File: python/179_Consecutive_positive_divisors.py
from functools import lru_cache
from math import isqrt

@lru_cache(None)
def n_divs(n):
    if n == 1:
        return 1
    if n == 2:
        return 2
    if n % 2 == 0:
        n //= 2
        c = 1
        while n % 2 == 0:
            c += 1
            n //= 2
        return (c + 1) * n_divs(n)
    for i in range(3,isqrt(n)+1,2):
        if n % i == 0:
            n //= i
            c = 1
            while n % i == 0:
                c += 1
                n //= i
            return (c + 1) * n_divs(n)
    return 2

def p179():
    last = 1
    n = 0
    for i in range(2,10000001):
        curr = n_divs(i)
        if curr == last:
            n += 1
        last = curr
    return n

print(p179())

File: c/179_Consecutive_positive_divisors.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

long n_divs(long n) {
  long c = 0;
  long isqrt = (long)sqrt(n);
  long i;
  for (i = 1; i <= isqrt; i++) {
    if (n % i == 0)
      c += 2;
  }
  if (n % (i - 1) == 0 && n / (i - 1) == (i - 1))
    --c;
  return c;
}

int main(int argc, char **argv) {
  long l = strtol(argv[1], NULL, 10);
  long past, curr;
  past = 1;
  long n = 0;
  for (long i = 2; i < l; i++) {
    curr = n_divs(i);
    if (curr == past) {
      ++n;
    }
    past = curr;
  }
  printf("%ld\n", n);
  return 0;
}
